tests/pass/shims/ctor.rs

```diff
@@ -0,0 +1,43 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+static COUNT: AtomicUsize = AtomicUsize::new(0);
+
+unsafe extern "C" fn ctor() {
+    COUNT.fetch_add(1, Ordering::Relaxed);
+}
+
+macro_rules! ctor {
+    ($ident:ident = $ctor:ident) => {
+        #[cfg_attr(
+            all(any(
+                target_os = "linux",
+                target_os = "android",
+                target_os = "dragonfly",
+                target_os = "freebsd",
+                target_os = "haiku",
+                target_os = "illumos",
+                target_os = "netbsd",
+                target_os = "openbsd",
+                target_os = "solaris",
+                target_os = "none",
+                target_family = "wasm",
+            )),
+            link_section = ".init_array"
+        )]
+        #[cfg_attr(windows, link_section = ".CRT$XCU")]
+        #[cfg_attr(
+            any(target_os = "macos", target_os = "ios"),
+            link_section = "__DATA,__mod_init_func"
```

Comment:
```suggestion
            link_section = "__DATA,__mod_init_func,mod_init_funcs"
```

Mach-O requires the `S_MOD_INIT_FUNC_POINTERS` section flag (`mod_init_funcs` in `link_section`). LLVM was accidentally adding this for `__DATA,__mod_init_func` making it work without explicitly adding `,mod_init_funcs`, but I don't think it guarantees this. See also https://github.com/rust-lang/rustc_codegen_cranelift/issues/1588

Reply:
From the [Mach-O reference](https://github.com/aidansteele/osx-abi-macho-file-format-reference), it looks like `mod_init_funcs` is not strictly required for the `S_MOD_INIT_FUNC_POINTERS` to be set.

> `S_MOD_INIT_FUNC_POINTERS`—This section contains pointers to module initialization functions.
The standard tools create `__DATA,__mod_init_func`sections of this type

Reply:
The docs seem to say that S_MOD_INIT_FUNC_POINTERS indicates init functions, which sounds to me like the flag is required to designate a section of init functions?

Reply:
If it is questionable whether the version without the extra tag *should* work (as opposed to "happens to work due to undocumented LLVM accidents"), then at the very least we need a FIXME comment here clarifying this.

> the ecosystem relies on it

Do you have a link for that?

Reply:
The [ctor](https://docs.rs/ctor/latest/src/ctor/macros/mod.rs.html#533) and [inventory](https://docs.rs/inventory/latest/src/inventory/lib.rs.html#550) crates do not set it explicitly.

I can add a comment to the code. I'm not even sure where to find the official Mach-O reference, all I can find are outdated mirrors.

Reply:
LLVM currently implies the flag for `__DATA,__mod_init_func`, but I don't know if it actually guarantees this. The implication is because it creates a section with this name with the flag set somewhere and then merges the attributes of all definitions of the same section.

Reply:
I have added a comment, that should suffice for now -- thanks for the links!